Runtime/JSMarshalAsAttribute.cs

```diff
@@ -0,0 +1,37 @@
+
+using System;
+
+namespace NodeApi;
+
+/// <summary>
+/// Specifies how .NET types are marshalled to and from JavaScript values.
+/// </summary>
+[AttributeUsage(
+    AttributeTargets.Property |
+    AttributeTargets.Parameter |
+    AttributeTargets.ReturnValue)]
+public sealed class JSMarshalAsAttribute : Attribute
+{
+    /// <summary>
+    /// Specifies how a .NET property, parameter, or return value is marshalled to
+    /// and from a JavaScript value.
+    /// </summary>
+    /// <param name="marshal">One of the available marshaling options.</param>
+    public JSMarshalAsAttribute(JSMarshal marshal)
+    {
+        Value = marshal;
+    }
+
+    /// <summary>
+    /// Gets the marshaling behavior indicated by the attribute.
+    /// </summary>
+    public JSMarshal Value { get; }
+
+    /// <summary>
+    /// When <see cref="Value" /> is <see cref="JSMarshal.CustomMarshaler" />, gets the type
+    /// that handles marshalling. The type must implement <see cref="IJSMarshaler{T}" />,
+    /// where the type argument matches the type being of the property, parameter, or
+    /// return value the <see cref="JSMarshalAsAttribute" /> is applied to.
+    /// </summary>
+    public Type? CustomMarshalType { get; init; }
```

Comment:
>Type? [](http://example.com/codeflow?start=11&length=5)

I wonder if we can use the new generic attributes instead. In that case the restriction that the type must inherit from IJSMarshaler can be written as an explicit restriction checked by the C# compiler.

Reply:
I thought about that, but we'd still need a non-generic `JSMarshalAsAttribute` to support uses that don't specify a custom marshaller type. So I didn't think it was worth creating another attribute.